providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -72,435 +40,441 @@ static OSSL_FUNC_keymgmt_import_types_fn mlkem_imexport_types;
-
-/* TODO(ML-KEM) template code to be completed as and when needed: */
-#ifdef UNDEF
-    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
-        int key_checked = 0;
-
-        if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
-            const uint8_t *pa = key1->pubkey;
-            const uint8_t *pb = key2->pubkey;
-
-            if (pa != NULL && pb != NULL) {
-                ok = ok
-                    && key1->keytype == key2->keytype
-                    && CRYPTO_memcmp(pa, pb, MLKEM768_PUBLICKEYBYTES) == 0;
-                key_checked = 1;
-            }
-        }
-        if (!key_checked
-            && (selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0) {
-            const uint8_t *pa = key1->privkey;
-            const uint8_t *pb = key2->privkey;
-
-            if (pa != NULL && pb != NULL) {
-                ok = ok
-                    && key1->keytype == key2->keytype
-                    && CRYPTO_memcmp(pa, pb, MLKEM768_SECRETKEYBYTES) == 0;
-                key_checked = 1;
-            }
-        }
-        ok = ok && key_checked;
+    /*
+     * Fail if the public key (or if requested the private key) is unavailable.
+     * The public key is available when either was provided.
+     */
+    if (!have_keys(key)
+        || ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0
+            && key->prvkey == NULL)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return 0;
     }
-#endif /* UNDEF */
-    debug_print("MLKEMKM match result %d\n", ok);
-    return ok;
+    return 1;
 }
 
-static int key_to_params(MLKEM768_KEY *key, OSSL_PARAM_BLD *tmpl,
-                         OSSL_PARAM params[], int include_private)
+static int mlkem_match(const void *vkey1, const void *vkey2, int selection)
 {
-    if (key == NULL)
-        return 0;
+    const MLKEM_KEY *key1 = vkey1;
+    const MLKEM_KEY *key2 = vkey2;
 
-    if (key->keytype != MLKEM_KEY_TYPE_768)
-        return 0;
-
-    if (key->encoded_pubkey != NULL
-        && !ossl_param_build_set_octet_string(tmpl, params,
-                                              OSSL_PKEY_PARAM_PUB_KEY,
-                                              key->encoded_pubkey,
-                                              OSSL_MLKEM768_PUBLIC_KEY_BYTES))
+    if (!ossl_prov_is_running())
         return 0;
 
-    if (include_private
-        && key->encoded_privkey != NULL
-        && !ossl_param_build_set_octet_string(tmpl, params,
-                                              OSSL_PKEY_PARAM_PRIV_KEY,
-                                              key->encoded_privkey,
-                                              OSSL_MLKEM768_PRIVATE_KEY_BYTES))
-        return 0;
+    /* All we have that can be compared is key material */
+    if (!(selection & OSSL_KEYMGMT_SELECT_KEYPAIR))
+        return 1;
 
-    return 1;
+    return ossl_mlkem_vcompare_pubkeys(key1->vinfo, key1->pubkey, key1->prvkey,
```

Comment:
Why just compare pubkeys? What if selection asks for privkeys to be matched? Or looking one level further at the contents of ossl_mlkem_vcompare_pubkeys (and if understanding the expansion behind "constish" right -- big if), should the function be rather re-named "ossl_mlkem_vcompare_keypairs"?

Reply:
I've beed told that when comparing keys it is in fact necessary to consider a private key as being the *same* key as its associate public key.